src/sinks/mqtt/sink.rs

```diff
@@ -0,0 +1,215 @@
+                event = input.next() => {
+                    let mut event = if let Some(event) = event {
+                        event
+                    } else {
+                        break;
+                    };
+
+                    let finalizers = event.take_finalizers();
+
+                    let topic = match self.connector.topic.render_string(&event) {
+                        Ok(topic) => topic,
+                        Err(error) => {
+                            emit!(TemplateRenderingError {
+                                error,
+                                field: Some("topic"),
+                                drop_event: true,
+                            });
+                            finalizers.update_status(EventStatus::Errored);
+                            continue;
+                        }
+                    };
+
+                    self.transformer.transform(&mut event);
+
+                    let event_byte_size = event.size_of();
+
+                    let mut bytes = BytesMut::new();
+                    let message = match self.encoder.encode(event, &mut bytes) {
+                        Ok(()) => {
+                            bytes.to_vec()
+                        }
+                        Err(_) => {
+                            finalizers.update_status(EventStatus::Errored);
+                            continue;
+                        }
+                    };
+                    let message_len = message.len();
+
+                    let qos = match self.quality_of_service {
+                        MqttQoS::AtLeastOnce => QoS::AtLeastOnce,
+                        MqttQoS::AtMostOnce => QoS::AtMostOnce,
+                        MqttQoS::ExactlyOnce => QoS::ExactlyOnce,
+                    };
+                    let retain = false;
+                    match client.publish(&topic, qos, retain, message).await {
+                        Ok(()) => {
+                            self.events_sent.emit(CountByteSize(1, event_byte_size));
+                            self.bytes_sent.emit(ByteSize(message_len));
+
+                            self.finalizers_queue.push_back(finalizers);
+                        }
+                        Err(error) => {
+                            emit!(MqttClientError { error });
+                            finalizers.update_status(EventStatus::Errored);
+                            return Err(());
+                        }
+                    }
+                },
+
+                else => break,
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl StreamSink<Event> for MqttSink {
+    async fn run(mut self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let input = input.fuse().peekable();
+        pin_mut!(input);
+
+        while input.as_mut().peek().await.is_some() {
+            let (client, connection) = self.connector.connect();
+            pin_mut!(client);
+            pin_mut!(connection);
```

Comment:
Does this need to be created inside the `while` loop?

Reply:
Good question. According to the current implementation, after handling the events, the connection is closed. I do not the original motivation of the author for this. @astro could you clarify, please?

IMO, would be better to move this logic outside of the loop, but maybe I do not know some MQTT brokers' specifics.

Reply:
If the connection is supposed to stay open it looks like the `connect()` call should really be outside the loop.

Reply:
Moved this outside of the loop.